### ISSUE TITLE
NEP 35: Finalize `like=` argument behaviour before 1.21 release

### ISSUE THREAD

**Issue by seberg** [numpy/numpy#17075]
Before we release 1.20, we should finalize the behaviour. Or possibly hide/replace the API. Unless NEP 35 was officially accepted, there are still alternative suggestions such as `np.get_array_module` which may win the race.

There are also a few issues which require attention if we decide to keep the API largely untouched:

1. (This is currently NOT valid anymore, so it is the strict version)  `l = [1, 2]; np.asarray(l, like=l)` is currently valid.  However, modern API may want to only accept array-object as input and not objects such as lists/tuples, which the user should convert ahead of time.  The above spelling could raise an error here, requiring the use of a different function (e.g. `np.asduckarray()`) to achieve the no-error version, which most existing API currently uses:

    ```python
    def strict_library(x):
         x = np.asarray(x, like=x)  # raise error if x is a list

    def non_strict_library(x):
         x = np.asduckarray(x)
    ```
    the alternative might be to force the use of a different API:
    ```python
    def strict_library(x):
         x = np.asarray(x, like=x, strict=True)
    ```
   (Note that there *might* be a use for supporting an `obj.__asduckarray__()` method to mirror `obj.__array__()`, which would require `np.asduckarray()`)

2. (This seems like an extension, since 1. is strict it is still possible) There may be a use for finding a common array-like for multiple input functions:
    ```python
    def function(arr1, arr2):
        tmp = np.arange(1000, like=np.common_arraylike(arr1, arr2))
    ```
    it was previously discussed that this could be achieved by making like work on tuples: `np.arange(1000, like=(arr1, arr2)`.

3. Ensure that the name (currently `like=`) is thought out and has been discussed sufficiently.

4. ~(seberg) Are we sure that some of these function should never dispatch based on the existing (usually scalar) input? (I guess so, but I thought I saw `np.linspace` in there, which made me wonder, but I suppose its not there).  In any case, we may want to review each function more individually.~  (Linspace is not include, so this seems fine)

5. (We currently dispatch aliases) Should we dispatch the aliases `np.asarray()`, etc. or rather only use the full `np.array()` call.  The reason for inclusion is currently that CuPy would have issues with the last version, because it [doesn't support the `subok` argument](https://docs.cupy.dev/en/stable/reference/generated/cupy.array.html) (there are no cupy subclasses):

**Comment by pentschev** [comment]
A few comments on the 4 points from the description:

1. This currently follows the same pattern of `__array_function__`, where dispatching on a `list` would return a `np.ndarray`, for example:

```python
>>> import numpy as np
>>> np.ones_like([1, 2, 3])
array([1, 1, 1])
>>> np.concatenate(([1, 2], (3, 4)))
array([1, 2, 3, 4])
```

I think changing that would go against the behavior in the rest of `__array_function__`.

2. I'm still not certain if this is a job for NumPy, maybe it would make more sense that the application handles that? For example, Dask may want to implement a `da.common_arraylike` and decide what to do if the user passes `(cupy_array, sparse_array)`, or any other special cases it wants to support, maybe?

3. The naming `like=` was mainly chosen and somewhat loosely agreed upon given its interchangeability with the name of `empty_like` and derived functions, where `like=` behaves similarly to the intent that `_like` as a suffix have in those functions.

4. I originally mentioned `np.linspace` in NEP-35 because I failed to notice that it [already dispatches on `__array_function__`](https://github.com/numpy/numpy/blob/64619c55896dad1927afe78c0702a58abf0d038f/numpy/core/function_base.py#L18-L25), that mention has been removed in https://github.com/numpy/numpy/pull/17093 .

**Comment by pentschev** [comment]
I've been thinking about this the past few days, and I see your point @rgommers in dealing with existing vs future libraries. Generally speaking, I understand the point and have no direct objections to requiring something like the following, as per your example:

```python
arr = np.asduckarray(arr)   # convert array_like's to numpy arrays
padding = np.asarray(padding, like=arr)
```

However, I don't really like that within Dask, for example, we would need probably everywhere to call `np.asduckarray` and pass the result to `like=`, which would probably overcomplicate things and eventually lead to many places where people will just forget the `np.asduckarray` call before `np.some_array_creation(..., like=arr)`. Granted, `like=` is also probably easy to forget as well, but requiring both is going to be burdensome for maintainers. The second point is we'll need to create that new array for all non-arrays, thus increasing the memory footprint for an array that is not even going to be used apart from verifying its type, which seems odd too.

A suboptimal solution would be to have two arguments, `like=` (for future libraries) and `duck_like=` (for existing libraries, where this would internally be forgiving to non-arrays), but I emphasize that is suboptimal and very likely undesired as well. Perhaps there's some alternative solution, but I can't think of one now.

**Comment by rgommers** [comment]
> However, I don't really like that within Dask, for example, we would need probably everywhere to call `np.asduckarray` and pass the result to `like=`,

My intuition says you're not having to add an extra `asduckarray` call, but are amending an existing check. But I'm not sure - can you point to some examples of code in Dask where you'd like to add `like=`?

**Comment by pentschev** [comment]
> > However, I don't really like that within Dask, for example, we would need probably everywhere to call `np.asduckarray` and pass the result to `like=`,
> 
> My intuition says you're not having to add an extra `asduckarray` call, but are amending an existing check. But I'm not sure - can you point to some examples of code in Dask where you'd like to add `like=`?

Certainly, for example https://github.com/dask/dask/blob/51509339c1cc476f5f96ae78e61daf125037ca20/dask/array/slicing.py#L878 . Note that it explicitly checks if it's an instance of `list` or `np.ndarray` in the line just above, which would probably require extending to `or hasattr(ind, "__array_function__")` or something similar.

**Comment by rgommers** [comment]
That code is pretty weird, I hope there's not too many explicit `isinstance` checks like that. I imagine you want to write it like
```
elif isinstance(ind, (list, np.ndarray) or hasattr(ind, "__array_function__")):
    x = np.asanyarray(ind, like=ind)
    if x.dtype == bool:
        ...
```
and avoid turning a duckarray that's not a subclass into a `numpy.ndarray` with `asanyarray`?

There's no reason to use `like=` at all there, looks to me like that code would be better written as:
```
if isinstance(ind, list);
    ind = np.asarray(ind)

if hasattr(ind, "__array_function__")):
    if x.dtype == bool:
       ...
```
It likely doesn't matter much, but that way it's also faster - avoids the `asanyarray` call for array input.

**Comment by pentschev** [comment]
Sorry @rgommers for the late reply, I'm still OOO until the end of this week and have been forcing myself on reducing screen time.

I agree that the example above is weird, and perhaps a better one would be in https://github.com/dask/dask/blob/51509339c1cc476f5f96ae78e61daf125037ca20/dask/array/slicing.py#L539-L542 , where we have `index` being an array (e.g., `cupy.ndarray`) and `chunks` as a `tuple`, and could rewrite it as follows:

```python
    index = np.asanyarray(index)
    cum_chunks = cached_cumsum(chunks)
    cum_chunks = np.asarray(cum_chunks, like=index)

    chunk_locations = np.searchsorted(cum_chunks, index, side="right")
```

**Comment by rgommers** [comment]
@pentschev no worries at all, reduced screen time sounds like an excellent idea.

> `index = np.asanyarray(index)`

That will still convert to `numpy.ndarray`, so that should be either
```
index = np.asanyarray(index, like=index)
```
or
```
# note: this won't quite do asanyarray for older numpy's according to NEP 30, so may need
# a tweak - maybe add a as_subclass=False keyword?
index = np.duckarray(index)
```

The first option is very weird, so I'd say the `duckarray` one is preferred.

Also there is no `cupy.searchsorted` yet, but once there is, your code looks about right. The extra `np.asarray` is indeed needed, to turn tuples into arrays rather than let `numpy.searchsorted` do that internally. However, it is needed either way, independent of the decision on the more strict `like=` that would raise an exception as I propose. So it's still not a counter-example of the strict mode being more burdensome for Dask.

**Comment by pentschev** [comment]
> > `index = np.asanyarray(index)`
> 
> That will still convert to `numpy.ndarray`, so that should be either
> 
> ```
> index = np.asanyarray(index, like=index)
> ```

Sorry @rgommers , indeed, what I meant was `index = np.asanyarray(index, like=index)`, I was focusing specifically on the `chunks` part and ignored `index`.

> Also there is no `cupy.searchsorted` yet, but once there is, your code looks about right.

I wasn't focusing on the downstream functionality either, but in this case it is implemented, see https://github.com/cupy/cupy/pull/2726, documentation was missing and got fixed in https://github.com/cupy/cupy/pull/3908 though.

> The extra `np.asarray` is indeed needed, to turn tuples into arrays rather than let `numpy.searchsorted` do that internally. However, it is needed either way, independent of the decision on the more strict `like=` that would raise an exception as I propose. So it's still not a counter-example of the strict mode being more burdensome for Dask.

I agree that wasn't the best example, I would have to look more carefully at all use cases in Dask to quantify how much of a burden that would be. Doing another quick search, one other example I can find that I think would match that is https://github.com/dask/dask/blob/377965addc6be168ad1c697999ded6f9abdf6cce/dask/array/core.py#L3964 . Consider for instance we decide to follow the same ordering as of `__array_function__`, as described in NEP-18:

> NumPy will gather implementations of `__array_function__` from all specified inputs and call them in order: subclasses before superclasses, and otherwise left to right. Note that in some edge cases involving subclasses, this differs slightly from the current behavior of Python.

If that's how we should handle this case, we would then rewrite that similar to the following:

```python
args = [np.asarray(a, like=args[0]) if isinstance(a, (list, tuple)) else a for a in args]`
```

The leftmost argument in that case would dominate the array type, and if it's `list` or `tuple`, then we would need to handle the `TypeError`, probably with `np.duckarray`, which would be a bit burdensome. Perhaps this case is a bit of stretch though as I don't know if left-to-right ordering makes the most sense for this particular case.

**Comment by pentschev** [comment]
Sorry for the silence here @rgommers and @seberg , but I think I'm now confident on picking this backup to finalize it for 1.20.

I did quite some testing of NEP-35 in Dask (see https://github.com/dask/dask/pull/6738) and it was very positive overall. I was able to cover pretty much everything, with a couple of exceptions (`random.choice` and some `linalg`, details in the PR) for unrelated reasons. With that said, I'm confident NEP-35 is able to push Dask's coverage of `__array_function__` forward, currently I'm not aware of any `__array_function__` limitations that's still holding Dask back.

I also wasn't able to find any cases where `list`s and `tuple`s would pose a problem with `like=`, nor make the code more complex due to that, at least not in the already existing Dask tests or the ones that I introduced. Given that, I would be equally supportive of either strict or non-strict usage for those types with `like=`.

Apart from that, I just opened https://github.com/numpy/numpy/pull/17678 . The PR suggests allowing the passage of `like=` downstream, refer to its description, and let's please try to keep the discussion about that particular change in the PR itself, so nobody who's involved with that misses any important information related to that in here.

Finally, I have no relevant updates to my comment in https://github.com/numpy/numpy/issues/17075#issuecomment-676766093 -- and follow-up discussion -- that address the original description of this issue. I see only that a new item (number 5) was added since my reply, addressing whether we should dispatch only `np.array` or aliases such as `np.asarray` too as we briefly discussed in https://github.com/numpy/numpy/pull/16935, and I still believe we should dispatch aliases as well, this is because downstream libraries are not required to implement such aliases as functions of `np.array` itself, some relevant comments begin in https://github.com/numpy/numpy/pull/16935#discussion_r477650879 .

**Comment by hameerabbasi** [comment]
> We could plan a meeting dedicated to these decisions end of next week or so, if it helps. We definitely have to call this experimental, and I am not quite sure what it would mean for dask if we need to modify behaviour.

A meeting with all interested parties seems like the best way to go about this.

**Comment by pentschev** [comment]
> We have to make a final decision soon for the 1.20 release. From my perspective we were tending towards the strict `like=` argument (it seems like [gh-17678](https://github.com/numpy/numpy/pull/17678) is unnecessary and superseded now)?

I think that seems to be a preference for you and @rgommers at least. From my perspective, either way will work, so I'll defer the final solution to both of you.

> My problem is that I think we need to do something, but I am not sure if this can be a final right solution. I don't really see the `like` argument being major hassle/addition API wise, it should be fairly easy to deprecate again (if annoying I admit) and it seems like a major improvement for `__array_function__` which is not going to go away in the midterm. And even if it is limited to the dask/cupy world, I guess the impact may effectively be pretty big, but I can't really judge it.

Indeed, this is a major improvement for Dask especially, it will finally allow us to have complete (or at least complete to the extent of what I've seen so far) coverage of `dask.array` with `__array_function__`.

> Dispatching aliases seems fine to me. Did we already have a decision on whether the tuple syntax is useful here?

I'm not sure I understand what you mean by tuple syntax, are you referring to things like `like=(arr1, arr2)`?

> We could plan a meeting dedicated to these decisions end of next week or so, if it helps. We definitely have to call this experimental, and I am not quite sure what it would mean for dask if we need to modify behaviour.

I'd be ok with that, if others feel that's necessary.

> _I am more than happy to do the code fixups, getting the decisions out of the way is the tricky part here._

I can also take care or help with that if we have a decision before December. I'll be moving in early-/mid-December, so my availability may be compromised then.

**Comment by rgommers** [comment]
> I think that seems to be a preference for you and @rgommers at least. From my perspective, either way will work, so I'll defer the final solution to both of you.

I think that is preferable (from memory), but unfortunately I have been really short on time recently. Given that I didn't/don't have the bandwidth to spend much time on this, I'll just say that I'd prefer to make that change but can also live with the alternative. Happy either way.

No need for a meeting I think. That does not solve a lack of time issue - on the contrary. Plus we already touched on this in the community meeting yesterday.

**Comment by eric-wieser** [comment]
Apologies for dropping in late here

> Saying that `like=` is implicit is odd, it's a _very_ explicit request. `like=instance_of_some_type` should give back an instance of `some_type`.

I think whether or not `like=some_list` makes sense comes down to whether you agree with this statement, and whether you interpret `like=some_cupy_array` as `"the result must be a cupy array"` or the weaker "the cupy dispatcher must be involved in deciding the result type". If you take it as the latter, then "the list dispatcher" can just be handwaved away as "the default coercion behavior".

**Comment by seberg** [comment]
@eric-wieser curious do you have a preference in any case? My tendency is still towards strict (and it seems like the more conservative start), so I am planning to merge that tomorrow when I wake up, but interested to thoughts.

I would argue it is probably "the result must be compatible with a cupy array", which is slightly weaker. But I like type safety, and if our vision for the future is to err on the type-safe side, an error seems right.  For example, a library function is passed an argument `array = [cupy_arr1, cupy_arr2]` can't really do the right thing unless it is strict.

**Comment by pentschev** [comment]
I'm not sure we can enforce "the result must be a CuPy array", the downstream implementation will return an array something based on the arguments passed to the function. A hypothetical case could be that `np.asarray([], like=downlib.array([1, 2, 3]))` dispatches to `downlib` and `downlib` decides that the appropriate return type for `downlib.asarray([])` is, for whatever reason, a `None` instead of `downlib.ndarray`. In that sense, I think we can only ensure "the CuPy dispatched must be involved in deciding the result". The only way I think we could enforce "the result must be a CuPy array" is to actually check the output and verify if it's an instance of `type(like)`, which IMO isn't NumPy's job, nor a case considered by `__array_function__` as a whole.

**Comment by charris** [comment]
Can this be closed now?

**Comment by eric-wieser** [comment]
> @eric-wieser curious do you have a preference in any case? My tendency is still towards strict (and it seems like the more conservative start)

My preference here is towards permissive, but I agree it's easy to make things more permissive later and hard to make them more conservative.

**Comment by seberg** [comment]
@charris I think we are OK for 1.20, I moved the milestone.  I am sure we should revisit it then, maybe even go with the permissive one.

**Comment by seberg** [comment]
This may be slightly annoying:
```
class myarr(np.ndarray): 
    pass 

arr = np.empty(5).view(myarr)  
res = np.asarray(3, like=arr)
```
Will not use `subok` and return a base-class ndarray.  It basically only works due to "optimizations" right now.  This also means that such a pattern, which used to work for subclasses is broken for the `like=` argument:
```
def __array_function__(self, func, types, args, kwargs):
    return super().__array_function__(func, types, args, kwargs)
```
although I doubt there was ever any much intention that the above really should work.

The reason why is this:
* `__array_function__` requires there to be a non-dispatching version of a function, currently stored as `public_api._implementation`
* The `like` argument on `np.array` must be defined in C, the python wrapper we use is just unacceptably slow.  But in C providing two implementations is annoying, and storing it as `public_api._implementation` would require custom callable objects.

So, the current solution is inconvenient for these C implemented functions.  `__array_ufunc__` solves this differently: It guarantees that the dispatching will never be performed on `ndarray.__array_function__` (which could be inherited). This removes the need for an `_implementation` (by removing the possibility of recursive calls).  It also disables the ability to use `return super().__array_ufunc__` (it should cause a recursive call).

Basically, the NumPy implementation *is* special, for one it has to avoid actually calling `ndarray.__array_function__` as that would be far too slow. So... the simplification gained by avoiding the dance `__array_ufunc__` does ends up being a bit of a complication since it adds the requirement of having a non-dispatching "internal" version of the function around (that requirement costs little in Python, but complicates things for C – and possibly some optimization schemes).

**Comment by seberg** [comment]
Ok, to be honest, the result of dropping the subclass is fine. What bothers me is probably mainly that it only works due to optimizations.  And right now the `__array_ufunc__` approach seems a bit easier from an implementation standpoint (i.e. more freedom within NumPy).

**Comment by pentschev** [comment]
Honestly, I don't see why this is a problem or annoying. As per both NEP-18 and NEP-35, we're dispatching to the downstream library, and the downstream library is responsible for deciding what's the appropriate behavior given its capabilities. For instance, neither CuPy nor Dask implement the `subok`, if we were to force that it wouldn't be possible to use `np.asarray` at all.

**Comment by seberg** [comment]
Yeah, behaviour wise it probably doesn't matter much.  The way `__array_ufunc__` works gives us more freedom. I.e. `np.array` is defined in C, but right now it would be very inconvenient to define `np.may_share_memory(a, b)` in C.

**Comment by charris** [comment]
ISTR this going by. What is the current status?

**Comment by seberg** [comment]
@pentschev do you think there is any usage we have to reconsider here, or must clarify.  There is a slippery slope to being de-facto not experimental api anymore unless we revisit this soon enough.  (Ideally by accepting the NEP as final.)

**Comment by pentschev** [comment]
So far I haven't really encountered anything we need to clarify or reconsider. Examples of real usage of NEP-35 can be found in Dask, particularly https://github.com/dask/dask/pull/6738 which introduces support for NEP-35 and several functions that were impossible to implement before that, as well as percentile support (https://github.com/dask/dask/pull/7162), lstsq/cholesky (https://github.com/dask/dask/pull/7563) and bincount slicing (https://github.com/dask/dask/pull/7391).

IMO at this time NEP-35 as it stands is already successful. As we approach NumPy 1.21 release I think it would be great to consider accepting it as final, unless something presents itself until then.

**Comment by seberg** [comment]
I guess the main volatility may have been around whether to relax the `like=` to allow non-array-likes. But, that is the one thing we can also just do without much concern I think.

**Comment by pentschev** [comment]
> I guess the main volatility may have been around whether to relax the `like=` to allow non-array-likes. But, that is the one thing we can also just do without much concern I think.

Apologies for the late reply. I agree we could adjust later, but so far even that I'm not sure would be necessary. Therefore, I'm +1 on just accepting NEP-35 as is at this time, and we can revisit that later if a real use case shows up in the future.
